### PR TITLE
Replace scalariform with scalafmt

### PR DESCRIPTION
**Pull request by mdedetrich**
This PR replaces scalariform with scalafmt as has been done in other lightbend/akka projects, some additional notes:

* Rather than having the code formatted every time after a `compile` is done, instead we a github action is added which checks that the codebase is formatted when a PR opened. The github action uses scalafmt-native which is very fast (typically checks an entire project in ~5-20 seconds and its also done concurrently from the main Scala build). I have opted for this solution because typically speaking  formatters which trigger on compile/git hook's/ etc etc tend to not play nicely with other tools (such as IDE's, editors). Note that this is in line with the other lightbend/akka projects that I have updated to use scalafmt
* The PR is separated into 3 commits, one for removing scalariform, one that adds scalafmt and one that actually applies it. If you want to change the configuration let me know and I can rebase the PR with the update configuration.

Resolves: https://github.com/lightbend/paradox/issues/515

### CODE REVIEW

File: core/src/main/scala/com/lightbend/paradox/markdown/Frontin.scala [hunk 1]
```diff
@@ -30,20 +30,20 @@ object Frontin {
...
+        source.close()
     apply(lines)
   }
 
   def apply(str: String): Frontin =
     str.linesWithSeparators.toList match {
       case Nil => Frontin(Map.empty[String, String], "")
       case x :: xs if separates(x) =>
-        xs span { !separates(_) } match {
+        xs span !separates(_) match {
```
Comment (mdedetrich) [inline core/src/main/scala/com/lightbend/paradox/markdown/Frontin.scala:46]: So I managed to find an instance where scalafmt changes the meaning of the program and actually causes code not to compile. Will create a ticket upstream.

Comment (mdedetrich) [reply core/src/main/scala/com/lightbend/paradox/markdown/Frontin.scala:46]: Issue created at https://github.com/scalameta/scalafmt/issues/3182

Comment (pvlugter) [reply core/src/main/scala/com/lightbend/paradox/markdown/Frontin.scala:46]: Looks like fix is coming already, but spelling this out fully so it's not broken for now would be good.

File: tests/src/test/scala/com/lightbend/paradox/markdown/example.scala [hunk 2]
```diff
@@ -18,9 +18,9 @@ package com.lightbend.paradox.markdown
 
 //#github-path-link
 object GithubPathLink {
-  //#github-neither-path-link
+  // #github-neither-path-link
```
[the remark this answers is not in the document]
Comment (pvlugter) [reply tests/src/test/scala/com/lightbend/paradox/markdown/example.scala:21]: That should be fine

File: .scalafmt.conf [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+maxColumn = 120
+docstrings.style = Asterisk
+assumeStandardLibraryStripMargin = true
+rewrite.rules = [RedundantBraces, RedundantParens, SortImports, SortModifiers]
+indentOperator.exemptScope = all
+align.tokens = [
+  {code = "%", owner = "Term.ApplyInfix"},
+  {code = "%%", owner = "Term.ApplyInfix"},
+  {code = ":=", owner = "Term.ApplyInfix"},
+  "="
```
Comment (pvlugter) [inline .scalafmt.conf:12]: So no alignment for cases?

```
{code = "=>", owner = "Case"}
```

With everything else aligned, may be good to have this too?

**Review by pvlugter — COMMENTED** [review]
Cool, thanks for organising this.

File: core/src/main/scala/com/lightbend/paradox/markdown/Frontin.scala (hunk 1 above)
Comment (mdedetrich) [reply core/src/main/scala/com/lightbend/paradox/markdown/Frontin.scala:46]: So I would recommend one out of 2 options.

1. Format the entire codebase and then remove, fix the above problem and then temporarily remove `RedundantBraces` until the next scalafmt release.
2. Just wait for the next scalafmt release. In the rare cases where scalafmt breaks code it typically doesn't take long for a new release.

File: .scalafmt.conf (hunk 3 above)
Comment (mdedetrich) [reply .scalafmt.conf:12]: Can you provide an example of current Scala code and what you expect the formatted to be? As far as I can see `case` is formatted but maybe I am misunderstanding?

File: core/src/main/scala/com/lightbend/paradox/markdown/Frontin.scala (hunk 1 above)
Comment (pvlugter) [reply core/src/main/scala/com/lightbend/paradox/markdown/Frontin.scala:46]: I was thinking just write it in a way that's both formatted and not broken. Either of these should be fine:

```scala
xs.span(!separates(_))
```

```scala
xs span { x => !separates(x) }
```

File: .scalafmt.conf (hunk 3 above)
Comment (pvlugter) [reply .scalafmt.conf:12]: I thought I was cases that weren't aligned.

File: core/src/main/scala/com/lightbend/paradox/markdown/Index.scala [hunk 4]
```diff
@@ -148,13 +151,13 @@ object Index {
...
-        // only check first children
-        case first :: _ => linkRef(first, level)
-        case _          => None
-      }
+      case other =>
+        other.getChildren.asScala.toList match {
+          // only check first children
+          case first :: _ => linkRef(first, level)
+          case _ => None
+        }
```
Comment (pvlugter) [inline core/src/main/scala/com/lightbend/paradox/markdown/Index.scala:159]: @mdedetrich example of `case` `=>` that was aligned before, but not any longer

File: core/src/main/scala/com/lightbend/paradox/markdown/LinkCapturer.scala [hunk 5]
```diff
@@ -65,8 +63,8 @@ class LinkCapturer {
 
   private val plugins = Writer.defaultPlugins(Writer.defaultDirectives.map(_.andThen {
     case ext: ExternalLinkDirective => new NodeOverridingDirective(ext)
-    case ref: RefDirective          => new NonRenderingDirective(ref)
-    case other                      => other
+    case ref: RefDirective => new NonRenderingDirective(ref)
+    case other => other
```
Comment (pvlugter) [inline core/src/main/scala/com/lightbend/paradox/markdown/LinkCapturer.scala:67]: Another example of case statements that are no longer `=>` aligned

File: .scalafmt.conf (hunk 3 above)
Comment (pvlugter) [reply .scalafmt.conf:12]: Have commented on a couple of examples where this was aligned before.

File: core/src/main/scala/com/lightbend/paradox/markdown/Frontin.scala (hunk 1 above)
Comment (mdedetrich) [reply core/src/main/scala/com/lightbend/paradox/markdown/Frontin.scala:46]: So I realized that `xs span separates` also works and is the most concise/aesthetically pleasing. Changing it now.

**Comment by mdedetrich** [comment]
So I updated `.scalamft.conf` to align `case` fat arrows but now tests are failing because of the previous issue I posted where in some comments the snippets have extra whitespace

**Comment by mdedetrich** [comment]
Okay so I realized what the issue is, the `SnippetDirectiveSpec` is testing the actual content of the source code at `com/lightbend/paradox/markdown/example.scala` which obviously breaks when you apply a different formatter such as scalafmt or even change the configuration.

Due to this I added another commit called `Apply fixes to SnipDirectiveSpec due to format` which applies fixes so the tests pass but it might be wiser to just add `com/lightbend/paradox/markdown/example.scala` to the `project.excludePaths` (see https://scalameta.org/scalafmt/docs/configuration.html#project) which would mean that this file will never get formatted and hence the possibility of subtly breaking the `SnippetDirectiveSpec` if the formatting configuration changes.

Thoughts?

**Comment by pvlugter** [comment]
> Due to this I added another commit called `Apply fixes to SnipDirectiveSpec due to format` which applies fixes so the tests pass

I think that's fine. Projects using paradox will often have formatting like this enabled as well, so having the same here, where some snippet label comments have whitespace added and others don't, seems a good test.

**Review by pvlugter — APPROVED** [review]
Thanks! Let's go ahead with this. We can always tweak things as needed later on.

**Pull request merged**